### PR TITLE
Added Proper Goalie Assignment in STP

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Most of the details for this can be seen in https://github.com/UBC-Thunderbots/Software/issues/745, but the gist of it is that this removes a hack we had to try to make sure that only the goalie robot was assigned to the goalie tactic and replaces it with proper logic in STP. 

Main changes are:
- added proper handling for goalie tactics in `stp.cpp` by adding a `isGoalieTactic` method to the `Tactic` class
- changed `assignRobotsToTactics` to not return anything, because it was already doing all the work in-place anyhow
- added a `GoalieTestTactic` for testing new functionality
- several unit tests

### Testing Done
Lots of unit testing.

### Length Justifications
Cleaned up a couple of things along the way, logic is fairly non-trivial, and tests for tactic assignment are unfortunately quite long.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/stp.cpp
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/stp.cpp]: We technically already have a case for this in the below if-else block. Because of how `tactics` return `nullptr` if they have no robot assigned, the logic below is ok, but I do not think it's handled in the best way. I'd recommend consolidating these checks at the top where you added the new check, and make sure to log there as well.

File: src/software/ai/hl/stp/stp.cpp
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/stp.cpp:142]: WIth the additional of goalie handling, this function has gotten very long. Possibly out of scope for this PR and more approriate for an stp cleanup, but this function should probably be broken up in to `assignGoalie` and `assignOtherRobots`. Your call if you want to address that now or now.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Overall looks fine, stp just may need some cleanup

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp:142]: I did quickly try this, but it's actually trickier because we do some removal of tactics/robots in the goalie section that are needed for the rest.... if you have some suggestions here I'm happy to make changes here!

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Sorry, confused what you're asking for here. 
- I'm not comfortable relying on logic in the `Tactic` class here. 
- Which case in the below if-else do you think covers this?
- "I'd recommend consolidating these checks at the top ": what are "these checks"?

File: src/software/world/robot_capabilities.cpp
Comment (jonathanlew) [inline src/software/world/robot_capabilities.cpp]: why was this file added?

File: src/software/ai/hl/stp/stp.cpp
Comment (jonathanlew) [inline src/software/ai/hl/stp/stp.cpp]: nit: 
```suggestion
    // one robot assigned per team
```

File: src/software/ai/hl/stp/stp.cpp
Comment (jonathanlew) [inline src/software/ai/hl/stp/stp.cpp]: can we combine this `if(goalie)` with the `if(goalie)` above?

File: src/software/ai/hl/stp/stp.cpp
Comment (jonathanlew) [inline src/software/ai/hl/stp/stp.cpp]: consider using `remove_if` https://en.cppreference.com/w/cpp/algorithm/remove

File: src/software/ai/hl/stp/stp.cpp
Comment (jonathanlew) [inline src/software/ai/hl/stp/stp.cpp]: Is there a case where `goalie==nullopt` is not a problem? Should we log a warning in the `else` case?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Good change overall. I'd like to echo Mat's comments about how long `assignRobotsToTactics` is

File: src/software/world/robot_capabilities.cpp
Comment (garethellis0) [reply src/software/world/robot_capabilities.cpp]: Whoops, was carry-over from another architecture I tried where the goalie became a robot capability. Removed.

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Changed to "    // one robot per team is assigned to be one", is that OK?

Comment (jonathanlew) [reply src/software/ai/hl/stp/stp.cpp]: It just needs to make sense with the previous line

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Yep, done!

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Rephrased again.

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Change this and removed most manual iterator usage usage above as well! Can't believe I forgot about this................

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Hmm, I don't think it's super unreasonable? Consider the case where we pull our goalie.

File: src/software/ai/hl/stp/stp.cpp
Comment (jonathanlew) [inline src/software/ai/hl/stp/stp.cpp:169]: Why are we erasing the return value of remove_if?

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp:169]: This is the [erase-remove idiom](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom), it's how you're supposed to use `std::remove_if`, and saves on memory moving.

File: src/software/ai/hl/stp/stp.cpp
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/stp.cpp:142]: Made #1109 so I can take a separate look after.

File: src/software/ai/hl/stp/stp.cpp
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/stp.cpp]: On line, 98, the returned action will be a `nullptr` if the tactic does not have a robot assigned. This means we will not enter the following `if` cases for the action and intent, and will avoid performing any logic until we reach the `else` case on line 120. That is what I mean by that this is already covered. If tactics have no robot assigned, ultimately no logic will be run.

I also agree that not handling it explicitly at the start is not the correct way to do things. Now we are "handling" the case where the tactic has no robot both in your new check on line 93, and in the `else` case on line 120. These are the two checks I'm referring to that should be consolidated. We should only perform a single explicit check for robot assignment, at the beginning of the logic (where you currently added the check on line 93).

Is that more clear?

Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Yup, removed the check on line 93 for now.

File: src/software/world/robot_capabilities.cpp
Comment (garethellis0) [reply src/software/world/robot_capabilities.cpp]: Ok, actually removed now.

File: src/software/ai/hl/stp/stp.cpp
Comment (jonathanlew) [reply src/software/ai/hl/stp/stp.cpp:169]: Ah I see. This is necessary for ordered data structures. Previously, I'd been using sets

File: src/software/world/robot.h
Comment (jonathanlew) [inline src/software/world/robot.h]: Todo?

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/ai/hl/stp/stp.cpp
Comment (EvanMorcom) [inline src/software/ai/hl/stp/stp.cpp]: Does this not include 1 additional Tactic?
The goalie tactics are removed, but friendlyTeam().numRobots() should return the team size including the goalie?

**Review by EvanMorcom — COMMENTED** [review]
LGTM :100:  
Pending the current comments

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Great catch! changed to `        tactics.resize(non_goalie_robots.size());`

File: src/software/world/robot.h
Comment (garethellis0) [reply src/software/world/robot.h]: Whoops, that was sloppy. Added tests for getting robotcapabilities whitelist and blacklist.

File: src/software/ai/hl/stp/stp.cpp
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/stp.cpp]: I don't see the change?

**Review by MathewMacDougall — APPROVED** [review]
LGTM pending Evan's comment.

**Review by jonathanlew — APPROVED** [review]
Thanks for addressing all my comments. LGTM pending Evan's comments :+1:

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp]: Could you double check?
Screenshot from 2019-12-29 13-34-00

Comment (MathewMacDougall) [reply src/software/ai/hl/stp/stp.cpp]: I see it now :+1:

**Pull request merged**
